backend/llm/openai.py

```diff
@@ -2,136 +2,119 @@
 from repository.chat.get_chat_history import get_chat_history
 from repository.chat.update_chat_history import update_chat_history
 from repository.chat.update_message_by_id import update_message_by_id
-from supabase import Client  # For interacting with Supabase database
-from supabase import create_client
+from supabase import Client, create_client
 from vectorstore.supabase import (
     CustomSupabaseVectorStore,
 )  # Custom class for handling vector storage with Supabase
 
+from .base import BaseBrainPicking
+from .prompts.CONDENSE_PROMPT import CONDENSE_QUESTION_PROMPT
+
 logger = get_logger(__name__)
 
 
-class BrainPicking(BaseModel):
+class OpenAIBrainPicking(BaseBrainPicking):
     """
     Main class for the Brain Picking functionality.
     It allows to initialize a Chat model, generate questions and retrieve answers using ConversationalRetrievalChain.
     """
 
     # Instantiate settings
-    settings = BrainSettings()
+    brain_settings = BrainSettings()
 
     # Default class attributes
-    llm_name: str = "gpt-3.5-turbo"
+    model_name: str = "gpt-3.5-turbo"
     temperature: float = 0.0
+    brain_id: str
     chat_id: str
     max_tokens: int = 256
-
-    # Storage
-    supabase_client: Client = None
-    vector_store: CustomSupabaseVectorStore = None
-
-    # Language models
-    embeddings: OpenAIEmbeddings = None
-    question_llm: LLM = None
-    doc_llm: LLM = None
-    question_generator: LLMChain = None
-    doc_chain: LLMChain = None
-    qa: ConversationalRetrievalChain = None
-
-    # Streaming
-    callback: AsyncIteratorCallbackHandler = None
     streaming: bool = False
 
-    class Config:
-        # Allowing arbitrary types for class validation
-        arbitrary_types_allowed = True
-
     def __init__(
         self,
-        model: str,
+        model_name: str,
         brain_id: str,
         temperature: float,
         chat_id: str,
         max_tokens: int,
         user_openai_api_key: str,
         streaming: bool = False,
-    ) -> "BrainPicking":
+    ):
         """
         Initialize the BrainPicking class by setting embeddings, supabase client, vector store, language model and chains.
-        :param model: Language model name to be used.
-        :param user_brain_idid: The brain id to be used for CustomSupabaseVectorStore.
-        :return: BrainPicking instance
+        :return: OpenAIBrainPicking instance
         """
         super().__init__(
-            model=model,
+            model_name=model_name,
             brain_id=brain_id,
             chat_id=chat_id,
             max_tokens=max_tokens,
             temperature=temperature,
             user_openai_api_key=user_openai_api_key,
+            streaming=streaming,
         )
-        # If user provided an API key, update the settings
-        if user_openai_api_key is not None:
-            self.settings.openai_api_key = user_openai_api_key
 
+        # Set the class attributes
         self.temperature = temperature
```

Comment:
Isn't this already set in the `__init__` of BaseBrainPicking during super().__init__(**data) ? Why do you need to set temperature, model_name, chat_id, brain_id and max_tokens again here ?

Reply:
Good spot